runtime/compiler/optimizer/abstractinterpreter/AbsInterpreter.cpp

```diff
@@ -0,0 +1,3072 @@
+         break;
+      }
+   }
+
+void TR::AbsBlockInterpreter::arrayStore(TR::DataType type)
+   {
+   TR::AbsStackMachineState* state = getState();
+
+   if (type.isDouble() || type.isInt64())
+      state->pop(); //dummy
+
+   TR::AbsValue* value = state->pop();
+   TR_ASSERT_FATAL(type == TR::Int8 || type == TR::Int16 ? value->getDataType() == TR::Int32 : value->getDataType() == type, "Unexpected type");
+
+   TR::AbsValue *index = state->pop();
+   TR_ASSERT_FATAL(index->getDataType() == TR::Int32, "Unexpected type");
+
+   TR::AbsValue *arrayRef = state->pop();
+   TR_ASSERT_FATAL(arrayRef->getDataType() == TR::Address, "Unexpected type");
+
+   //heap is being not modeled
+   switch (type)
+      {
+      case TR::Double:
+      case TR::Float:
+      case TR::Int8:
+      case TR::Int16:
+      case TR::Int32:
+      case TR::Int64:
+      case TR::Address:
+         break;
+      default:
+         TR_ASSERT_FATAL(false, "Invalid type");
+         break;
+      }
+   }
+
+void TR::AbsBlockInterpreter::binaryOperation(TR::DataType type, BinaryOperator op)
+   {
+   TR::AbsStackMachineState* state = getState();
+
+   if (type.isDouble() || type.isInt64())
+      state->pop(); //dummy
+
+   TR::AbsValue* value2 = state->pop();
+   TR_ASSERT_FATAL(value2->getDataType() == type, "Unexpected type");
+
+   if (type.isDouble() || type.isInt64())
+      state->pop(); //dummy
+
+   TR::AbsValue* value1 = state->pop();
+   TR_ASSERT_FATAL(value1->getDataType() == type, "Unexpected type");
+
+   switch (type)
+      {
+      // float and double are not modeled
+      case TR::Float:
+         {
+         TR::AbsValue *result = createTopFloat();
+         state->push(result);
+         break;
+         }
+
+      case TR::Double:
+         {
+         TR::AbsValue *result1 = createTopDouble();
+         TR::AbsValue *result2 = createTopDouble();
+         state->push(result1);
+         state->push(result2);
+         break;
+         }
+
+      //The following types are modeled
+      case TR::Int32:
+         {
+         if (isIntConst(value1) && isIntConst(value2) && !insideLoop()) //both int const
+            {
+            int32_t intVal1 = static_cast<TR::AbsVPValue*>(value1)->getConstraint()->asIntConst()->getInt();
+            int32_t intVal2 = static_cast<TR::AbsVPValue*>(value2)->getConstraint()->asIntConst()->getInt();
+
+            if (intVal2 == 0 && op == BinaryOperator::div) //divide by zero exception
+               {
+               state->push(createTopInt());
+               break;
+               }
+
+            int32_t resultVal;
+
+            switch (op)
+               {
+               case BinaryOperator::plus:
+                  resultVal = intVal1 + intVal2;
+                  break;
+               case BinaryOperator::minus:
+                  resultVal = intVal1 - intVal2;
+                  break;
+
+               case BinaryOperator::mul:
+                  resultVal = intVal1 * intVal2;
```

Comment:
I added this trick for now, but I have a concern for casting an unsigned int to signed int, since if that unsigned int is out of the range of signed int, the behavior seems to be implementation-defined. 

For this case, it could be `intVal1 = INT32_MAX` and `intVal2 = 1`, and then the result of `(uint32_t)intVal1 + (uint32_t)intVal2` will be cast to signed int, which is a implementation-defined behavior.

Could you please confirm this kind of "implementation-defined behavior" is acceptable?

> 
    C++11(ISO/IEC 14882:2011) §4.7 Integral conversions [conv.integral/2]

    If the destination type is unsigned, the resulting value is the least unsigned integer congruent to the source integer (modulo 2^n where n is the number of bits used to represent the unsigned type). [ Note: In a two’s complement representation, this conversion is conceptual and there is no change in the bit pattern (if there is no truncation). —end note ]

    If the destination type is signed, the value is unchanged if it can be represented in the destination type (and bit-field width); otherwise, the value is implementation-defined.

Reply:
> First of all, I would like to double check that we would like to keep all `resultVal` being correctly wrapped and also need to avoid saturation when the it overflows, right?

Yes

> Could you please confirm this kind of "implementation-defined behavior" is acceptable?

Yes, it's OK. All of our platforms and build compilers use two's complement, and I think that we assume pretty widely that conversion between signed and unsigned integers of the same size preserves the bit pattern, i.e. that it selects the only value that's equivalent to the original value mod the appropriate power of two

Here are the documented implementation-defined behaviours I was able to find:

[GCC](https://gcc.gnu.org/onlinedocs/gcc/Integers-implementation.html):
> *The result of, or the signal raised by, converting an integer to a signed integer type when the value cannot be represented in an object of that type (C90 6.2.1.2, C99 and C11 6.3.1.3).*
>
> For conversion to a type of width *N*, the value is reduced modulo *2^N* to be within range of the type; no signal is raised.

[MSVC](https://learn.microsoft.com/en-us/cpp/c-language/demotion-of-integers?view=msvc-170):
> When `signed` variables are converted to `unsigned` and vice-versa, the bit patterns remain the same. For example, casting -2 (0xFE) to an `unsigned` value yields 254 (also 0xFE).

[XL C/C++ for z/OS](https://www.ibm.com/docs/en/zos/3.1.0?topic=reference-implementation-defined-behavior)
> The result of converting an unsigned integer to a signed integer of equal length, if the value cannot be represented:
>
> - The bit pattern is preserved and the sign bit has no significance.

Note also that C++20 has mandated two's complement, and more specifically it has mandated that these conversions behave in the way assumed here. [cppreference](https://en.cppreference.com/w/cpp/language/implicit_conversion):
> If the destination type is signed, the value does not change if the source integer can be represented in the destination type. Otherwise the result is
> - implementation-defined *(until C++20)*
> - the unique value of the destination type equal to the source value modulo *2^n* where *n* is the number of bits used to represent the destination type *(since C++20)*

This isn't to say that we can rely directly on the C++20 standard - to the contrary, we can't even use all of C++11 - but I believe that the standard has been updated only because essentially all compilers already implement the conversions this way

Reply:
Thank you Devin! I will use this as the guide for the other overflow issues.